## Project - Page Visits Funnel

Cool T-Shirts Inc. has asked you to analyze data on visits to their website. Your job is to build a funnel, which is a description of how many people continue to the next step of a multi-step process.

In this case, our funnel is going to describe the following process:

1. A user visits CoolTShirts.com
2. A user adds a t-shirt to their cart
3. A user clicks “checkout”
4. A user actually purchases a t-shirt

In [47]:
import pandas as pd

In [48]:
visits = pd.read_csv('visits.csv', parse_dates=[1])
cart = pd.read_csv('cart.csv', parse_dates=[1])
checkout = pd.read_csv('checkout.csv', parse_dates=[1])
purchase = pd.read_csv('purchase.csv', parse_dates=[1])

### Inspect the DataFrames using print and head:

visits lists all of the users who have visited the website
cart lists all of the users who have added a t-shirt to their cart
checkout lists all of the users who have started the checkout
purchase lists all of the users who have purchased a t-shirt

In [49]:
visits.head()

user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00

In [50]:
cart.head()

user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00

In [51]:
checkout.head()

user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 2017-04-20 16:15:00
4  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:52:00

In [52]:
purchase.head()

user_id       purchase_time
0  4b44ace4-2721-47a0-b24b-15fbfa2abf85 2017-05-11 04:25:00
1  02e684ae-a448-408f-a9ff-dcb4a5c99aac 2017-09-05 08:45:00
2  4b4bc391-749e-4b90-ab8f-4f6e3c84d6dc 2017-11-20 20:49:00
3  a5dbb25f-3c36-4103-9030-9f7c6241cd8d 2017-01-22 15:18:00
4  46a3186d-7f5a-4ab9-87af-84d05bfd4867 2017-06-11 11:32:00

### Combine visits and cart using a left merge.

In [53]:
v_c_merged = pd.merge(visits, cart, how='left')
v_c_merged.head()

user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
4  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3 2017-08-20 08:49:00  
4                 NaT

### How long is your merged DataFrame?

In [54]:
v_c_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2052 entries, 0 to 2051
Data columns (total 3 columns):
user_id       2052 non-null object
visit_time    2052 non-null datetime64[ns]
cart_time     400 non-null datetime64[ns]
dtypes: datetime64[ns](2), object(1)
memory usage: 64.1+ KB


### How many of the timestamps are null for the column cart_time? What do these null rows mean?

In [55]:
null_values = v_c_merged.cart_time.isnull().sum() # 2052 - 400 = 1652 so this is correct
null_values
# It means there is no value entered - nan - not a number

1652

### What percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart?

Note: To calculate percentages, it will be helpful to turn either the numerator or the denominator into a float, by using float(), with the number to convert passed in as input. Otherwise, Python will use integer division, which truncates decimal points.

In [56]:
v_c_merged.user_id.count()

2052

In [57]:
float(null_values / v_c_merged.user_id.count()) * 100

80.50682261208577

## Answer - 80.51%

### Repeat the left merge for cart and checkout and count null values. What percentage of users put items in their cart, but did not proceed to checkout?

In [58]:
c_c_merged = pd.merge(cart, checkout, how='left')
c_c_merged.head()

user_id           cart_time  \
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
1  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
2  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
3  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00   
4  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00   

        checkout_time  
0 2017-11-07 21:14:00  
1 2017-11-07 20:50:00  
2 2017-11-07 21:11:00  
3                 NaT  
4 2017-03-04 11:04:00

In [59]:
c_c_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 602 entries, 0 to 601
Data columns (total 3 columns):
user_id          602 non-null object
cart_time        602 non-null datetime64[ns]
checkout_time    476 non-null datetime64[ns]
dtypes: datetime64[ns](2), object(1)
memory usage: 18.8+ KB


In [60]:
c_c_null = c_c_merged.checkout_time.isnull().sum() # 602 - 476 = 126 so this is correct
c_c_null

126

In [61]:
float(c_c_null / c_c_merged.cart_time.count()) * 100

20.930232558139537

## Answer -  20.9%

### Merge all four steps of the funnel, in order, using a series of left merges. Save the results to the variable all_data.

Examine the result using print and head.

In [63]:
all_data = visits.merge(cart, how='left').merge(checkout, how='left').merge(purchase, how='left')
all_data.head()

user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
4  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   

            cart_time       checkout_time       purchase_time  
0                 NaT                 NaT                 NaT  
1 2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  
2 2017-08-20 08:31:00                 NaT                 NaT  
3 2017-08-20 08:49:00                 NaT                 NaT  
4                 NaT                 NaT                 NaT

### What percentage of users proceeded to checkout, but did not purchase a t-shirt?

In [70]:
checkout_purchase = pd.merge(checkout, purchase, how='left')
checkout_purchase_rows = len(checkout_purchase)
null_purchase_times = len(checkout_purchase[checkout_purchase.purchase_time.isnull()])
float(null_purchase_times / checkout_purchase_rows) *100

16.88963210702341

## Answer = 16.89%

### Which step of the funnel is weakest (i.e., has the highest percentage of users not completing it)?

How might Cool T-Shirts Inc. change their website to fix this problem?

## Visits

### Using the giant merged DataFrame all_data that you created, let’s calculate the average time from initial visit to final purchase. Start by adding the following column to your DataFrame:

```all_data['time_to_purchase'] = all_data.purchase_time - all_data.visit_time```

In [71]:
all_data['time_to_purchase'] = all_data.purchase_time - all_data.visit_time

Examine the results using:

```print(all_data.time_to_purchase)```

In [75]:
all_data.time_to_purchase.head(20)

0         NaT
1    00:44:00
2         NaT
3         NaT
4         NaT
5         NaT
6         NaT
7         NaT
8         NaT
9         NaT
10        NaT
11        NaT
12        NaT
13        NaT
14        NaT
15   00:38:00
16        NaT
17        NaT
18        NaT
19        NaT
Name: time_to_purchase, dtype: timedelta64[ns]

### Calculate the average time to purchase using the following code:

```print(all_data.time_to_purchase.mean())```

In [73]:
all_data.time_to_purchase.mean()

Timedelta('0 days 00:44:02.672413')